Feature/vnq 10366 clustering

alexander-nsk opened the pull request:
1) Правки по ревью
2) Отдельные активности
3) Мин. функционал по кластерам.
4) Обновление зависимостей gradle до последних версий.

alexander-nsk committed:
VNQ-10366: Review fixes.

alexander-nsk committed:
VNQ-10366: Updated gradle dependencies to the last version.

alexander-nsk committed:
VNQ-10366: Reformat JS code. Updated id and interface for Clusterer.

alexander-nsk committed:
VNQ-10366: Updated Clusterer id generator.

alexander-nsk committed:
VNQ-10366: Added separate activities for testing Markers and Clustering. Added some coordinates to test Clustering.

t-artikov commented on app/src/main/java/ru/dublgis/dgismobile/sdktestapp/ClusterersActivity.kt:
```diff
@@ -0,0 +1,150 @@
+import ru.dublgis.dgismobile.mapsdk.clustering.ClustererStyle
+import ru.dublgis.dgismobile.mapsdk.clustering.InputMarker
+
+class ClusterersActivity : AppCompatActivity() {
```
Значительная часть кода активитей одинаковая - управление картой, геолокация. Хорошо было бы избавиться от дублирования, наверно, уже в следующий раз, когда появятся еще примеры.

t-artikov commented on mapsdk/src/main/java/ru/dublgis/dgismobile/mapsdk/PlatformBridge.kt:
```diff
@@ -51,7 +53,7 @@ internal class PlatformBridge(
     private var _controls: Boolean = false
 
-    private var selectedIds = ArrayList<String>()
+    private var clusterId = Random.nextInt().absoluteValue
```
В чем смысл этого начального значения? Почему не начинать id-ы c нуля?

t-artikov commented on mapsdk/src/main/java/ru/dublgis/dgismobile/mapsdk/clustering/ClustererStyle.kt:
```diff
@@ -0,0 +1,19 @@
+typealias LabelAnchor = Pair<Double, Double>
+typealias Size = Pair<Double, Double>
+
+class ClustererStyle(
```
Из всех параметров создания `Clusterer`-а используется лишь `radius`.  Не стоит добавлять параметры "про запас", нужно либо сразу реализовывать либо пока что не добавлять вовсе. Я, как пользователь SDK, удивился бы, если меняю значения, а ничего не происходит.

t-artikov commented on mapsdk/src/main/java/ru/dublgis/dgismobile/mapsdk/clustering/ClustererImpl.kt:
```diff
@@ -2,26 +2,26 @@ package ru.dublgis.dgismobile.mapsdk.clustering
+        controller.get()?.loadClustererMarkers(id, clusterMarkers)
     }
 
     override fun setOnClickListener(listener: ClustererClickCallback?) {
```
Проверял, что обработка кликов работает?

t-artikov commented on mapsdk/src/main/java/ru/dublgis/dgismobile/mapsdk/clustering/ClustererOptions.kt:
```diff
@@ -1,6 +1,6 @@
 class ClustererOptions(
-    var clusterStyle: ClusterStyle,
+    var clustererStyle: ClustererStyle,
     var radius: Int
```
Кажется, радиус должен быть вещественным, а не целочисленным.

t-artikov commented on mapsdk/src/main/java/ru/dublgis/dgismobile/mapsdk/PlatformBridge.kt:
```diff
@@ -200,15 +205,15 @@ internal class PlatformBridge(
         jsExecutor(
             """
-            window.dgismap.addClusterMarkers($id, $arg);
+            window.dgismap.loadMarkers($id, $arg);
```
Лучше было оставить в названии метода, что это относится к кластерам - иначе кажется, что это обычные маркеры.

t-artikov commented on mapsdk/src/main/java/ru/dublgis/dgismobile/mapsdk/clustering/ClustererImpl.kt:
```diff
@@ -2,26 +2,26 @@ package ru.dublgis.dgismobile.mapsdk.clustering
 import ru.dublgis.dgismobile.mapsdk.PlatformBridge
 import java.lang.ref.WeakReference
+import java.util.*
 import kotlin.math.absoluteValue
```
Проверяй код с помощью Inspect Code в андроид студии, здесь, например, неиспользуемые импорты.

t-artikov commented on mapsdk/src/main/java/ru/dublgis/dgismobile/mapsdk/PlatformBridge.kt:
```diff
@@ -175,19 +177,22 @@ internal class PlatformBridge(
     }
 
-    override fun addClusterMarkers(id: String, inputMarkers: Collection<InputMarker>) {
+    fun loadClustererMarkers(id: String, inputMarkers: Collection<InputMarker>) {
```
Все параметры `InputMarker` кроме `position` не имею эффекта, лучше было не добавлять их вовсе.

alexander-nsk committed:
VNQ-10366:
1) Reformat the code and optimize imports
2) Remove ClustererStyle.kt and style params in InputMarker.kt
3) Create parent class for examples
4) Added getId method for Clusterer.kt

t-artikov commented on mapsdk/src/main/java/ru/dublgis/dgismobile/mapsdk/clustering/Clusterer.kt:
```diff
@@ -3,9 +3,11 @@ package ru.dublgis.dgismobile.mapsdk.clustering
 
 interface Clusterer {
-    fun show(clusterMarkers: Collection<InputMarker>)
+    fun getId(): String
```
Не уверен в необходимости этого поля, что пользователю с ним делать?

alexander-nsk committed:
VNQ-10366: Removed getId method for Clusterer.kt

alexander-nsk committed:
Merge branch 'master' into feature/VNQ-10366-clustering
